src/dmd/errors.d

```diff
@@ -400,6 +400,16 @@ extern (C++) void message(const(char)* format, ...)
     va_end(ap);
 }
 
+/// The type of the diagnostic handler
+alias DiagnosticHandler = void delegate(const ref Loc location, Color headerColor, const(char)* header, const(char)* messageFormat, va_list args, const(char)* prefix1, const(char)* prefix2);
+
+/**
+ * The diagnostic handler.
+ * This will be called for every diagnostic message issued by the compiler.
+ * By default it will print the location and the message to stderr.
+ */
+DiagnosticHandler diagnosticHandler;
```

Comment:
Usually there are no thread locals in dmd, but I think it is fine here to be usable in the frontend from multiple threads. (That will need more work on other globals, though). Or should we defer thread-local storage until other places are adapted, too?

Reply:
Let’s defer TLS.